feat: add `toConfig`

satazor opened the pull request:
The `toConfig` feature require some rethougts on some things:

- Ideally, `toConfig` wouldn't need to be re-implemented everytime a chain has child chains
- Because of that, `set` was changed to add a getter/setter if value is a chainable
- To make things more consistent, `extend` was renamed to `shorthands`
- Additionally, to be closer to the native Map, `entries` was changed to return an array of key value pairs
- Added `asArray` option to ChainedMap and OrderedChainedMap

Closes #2.

satazor commented on README.md:
```diff
@@ -184,13 +210,13 @@ class Jest extends ChainedMap {
         ]);
-
-        this.coverageThresholds = new ChainedMap(this);
+        this.chains({
```
I wonder if we could simplify this a bit and make it single:

```js
this.chain('coverageThresholds', new ChainedMap(this));
```

I made it accept multiple to be consistent with `shorthands`, but it doesn't need to be. Thoughts?

satazor replied:
I actually made the change and it's a bit more clean imho.

andrefgneves commented:
Should we consider not adding a `chain` method and define a property in every `set` call? It would reduce the API surface and cognitive load :) Is there some other gain I'm overlooking?

Also, I find it a bit confusing to have both a `toObject` and a `toConfig`. It's not immediately clear which one does what. Should we consider only having one of them (maybe just `toConfig`/`toPlain`) with a `bool` specifying whether or not to do a deep transform?

satazor commented:
@andrefgneves I've updated the PR, describing the sugar that `chain` adds and why relying in using `set` would be much more verbose. Specially now, that `chain` accepts a optional `toConfig` argument which is also exemplified in the README, see: https://github.com/moxystudio/js-chained-config/blob/522a190f5e952d4e23c04b7456d7a5135afd6170/README.md#example-usage-of
If you mean using `set` and detect the value type to do the getter/setter automatically, I guess it’s feasible. Thought, we must add a third argument to it which is an optional `toConfig` just like the current `chain` method. This perhaps makes things a bit messy.

Regarding `toConfig` vs `toObject`, the difference is that `toConfig` will call `toConfig` recursively while `toObject` does not. The new `toObject` was the old `entries` method, which I have changed to be consistent with native Map.

We can remove `toObject` because it's easier to do it anyways.. it's just `.entries()` + a reduce. Having `toConfig` accepting a bool would be kind of weird. Accepting a bool in `toObject` would be plausible but `toConfig` is much more semantic.

If you have alternative proposals for these two points, could you elaborate a bit with a few examples?

andrefgneves commented:
Can you elaborate on how defining the `setters/getters` in `set` would be more verbose?

`set` would become:

```js
 set(key, value) {
    this.store.set(key, value);

    // maybe check if this hasn't been done before
    Object.defineProperty(this, key, {
        get: () => this.get(key),
        set: (value) => this.store.set(key, value),
        enumerable: true,
        configurable: true,
    });

    return this;
}
```

and "chaining" would just be a regular `set`call:

```js
class JestConfig extends ChainedMap {
    constructor(parent) {
        super(parent);

        this.shorthands(['rootDir', 'runner']);

        this.set('coverageThresholds', new ChainedMap(this));
    }
}
```

and later on:

```js
const jest = new JestConfig();

jest.coverageThresholds.set('global', {});
```

Regardless of whether or not we keep `chain`, I think the `toConfig` override should be defined in the `ChainedMap` constructor:

```js
// some file

export default new ChainedMap(this, { toConfig: myCustomToConfig });

// some other file

import fooConfig from 'someFile';

this.set('coverageThresholds', fooConfig);
// or
this.chain('coverageThresholds', new ChainedMap(this, { toConfig: ... } });
```

This way, the `toConfig` is a part of the map instance and available directly on it.

---

About `toObject/toConfig`, I vote we remove `toObject`.

satazor commented:
> If you mean using set and detect the value type to do the getter/setter automatically, I guess it’s feasible

This is what I was expecting and I like the idea. I will update this PR.

Regarding `toObject`, I will remove it.

satazor commented:
Regarding `shorthands` vs `extends`, which one you prefer? I mainly made this change because we were departing from webpack-chain compat.

andrefgneves commented:
I prefer `shorthands`.

satazor commented:
Regarding the `toConfig` option, I think it should have the following signature:

```js
new ChainedMap(parent, {
    toConfig: (config, chainedMap) => Object.values(config)
});
```

The `config` argument would be what would be returned by default. The thing is that it's already cleaned,  that is, have empty arrays/object and undefined values stripped out. This is useful.. if we weren't doing this, we would have to do the cleanup ourselves. In case you want something different you have the second argument to call methods on and return the appropriate data.

Having that said, do you agree with it? If so, do you think that the option still should be named  `toConfig` or something else?

andrefgneves commented:
I'm ok with `toConfig`, but I agree that is feels funny in the `ChainedMap` context.

`toJSON`?

satazor commented:
Another approach is:

```js
new ChainedMap(parent, {
    toConfig: (toConfig, chainedMap) => Object.values(toConfig())
});
```

This way it works well with the `toConfig` name, with the benefit that we aren't pre-computing the `config` if it's not used. With this approach, the `toConfig` option means that we are wrapping `toConfig` and changing its behavior (like lodash's wrap).

Regarding `toJSON`, I prefer `toConfig` because of their semantic meaning. The issue with `JSON` is that in my mind, json is a string.. which actually isn't what we are returning.

andrefgneves commented:
btw, my rationale for `toJSON` was having something like this: https://developer.mozilla.org/en-US/docs/Web/API/Body/json

But `toConfig` is fine :)

satazor commented:
@andrefgneves I've actually simplified the `toConfig` option. What we want is to simply make `toConfig` return an array in some situations for ChainedMap and OrderableChainedMap. Having that said, I simply added a `asArray` option instead which is simpler. What do you think?

Regarding `toConfig` and `toJson`. I think I'm going change to `toJson([recurse=true])`!

andrefgneves commented on lib/Chainable.js:
```diff
@@ -1,8 +1,16 @@
+        Object.defineProperty(this, IDENTIFIER, { value: true });
+    }
+
+    static isChainable(value) {
```
Should we drink our own Kool-aid and use [js-class-is](https://github.com/moxystudio/js-class-is) for this? 😉

satazor replied:
Ahaha will do.

satazor replied:
@fsdiogo when will we be landing 1.0.0?

satazor replied:
I've discovered a bug in `js-class-is`. Made a PR:  https://github.com/moxystudio/js-class-is/pull/6

satazor replied:
The MR was merged! I updated the version and tests are now passing.

satazor commented:
Actually, `toJson` with recurse disabled doesn't make sense because it's no longer a json since the result might contain values that are not valid json types. It's even worse because they have circular references haha. Will keep toConfig for now.

andrefgneves commented:
I think the `asArray` option is to specific and limiting. What if I want, for example, to have a map with camel cased keys when using it, but transform them to snake case when exporting?

I like the previous approach of "wrapping" `toConfig`.

satazor commented:
@andrefgneves In that case, you may implement a class that extends `ChainedMap` and override `toConfig`. Makes sense?

andrefgneves commented:
Yeah, sure.

What use cases do you see for `asArray`?

satazor commented:
It's exemplified on the README but I explain here:

We can use ChainedMap as an "array".. it has a huge advantage in which we can delete or modify entries via a  "label" (key). More so with OrderableChainedMap where we can also order them via "labels".

The `asArray` allows us to use those constructors but have `toConfig` always return an array.

Makes sense?

andrefgneves commented:
Got it.

I would also mention accessing items by a label as a benefit in the Readme.
